File: app/javascript/game/Round.js
import * as THREE from "three"
import { euro } from "game/Tuning"

// The round as the server tells it: the town, the parade route, the obstacles and their state, the vote on the
// next town, the server clock offset and the shared action cooldown, plus every Dutch string on the HUD. Messages
// handled: sync (on subscribe), round (every status change), object (hit points), vote (the tally), teleport/switch
// (verdicts on actions), end. Hooks: onRound(body, { fresh, started, live }), onEnd(msg), onAction(msg) for the
// player's own accepted actions, onObjects(list) for the destructibles, onVote(vote or null).
const SLOTS = 60                                                              // stretches of route on the bar
const ICON = { m: "🏠", b: "🏠", t: "🌳", l: "💡", g: "🚦", s: "🪧", d: "🧱" }
const RUBBLE_ICON = "🧱"                                                      // a flattened thing still in the way
// the strip of road the float needs clear, and how far in front of it rubble is worth reporting at all. Both match
// Game::Round::DEBRIS_AHEAD on the server, which is the side that decides.
const PARADE = { corridor: 7, ahead: 45 }

export class Round {
  constructor(playerId, els, hooks) {
    this.playerId = playerId
    this.els = els
    this.hooks = hooks
    this.offset = 0                 // server clock minus Date.now()
    this.round = null               // the round body from the server, or null while idle
    this.obstacles = new Map()      // key → obstacle, states kept current from `object` messages
    this.vote = null                // the vote between rounds: { candidates, by, ends_at }, or null
    this.nextActionAt = null        // server ms; null = the action is ready
    this.damage = 0                 // euros of property the room has flattened this round (lib/game/round.rb totals it)
    this.flashTimer = null
  }

  now() { return Date.now() + this.offset }
  get status() { return this.round?.status }
  get running() { return this.status === "running" }

  receive(msg) {
    if (msg.now) this.offset = msg.now - Date.now()
    switch (msg.type) {
      case "sync":   this.nextActionAt = msg.you.next_action_at; this.setRound(msg.round, false); this.setVote(msg.vote ?? null); break
      case "round":  this.setVote(null); this.setRound(msg.round, true); break
      case "vote":
        if (msg.ok === false) this.flash(msg.reason === "unknown" ? "Die plaats ken ik niet" : "De stemming is gesloten")
        else this.setVote(msg.vote)
        break
      case "object": this.applyObjects(msg.list, msg.damage); break
      case "end":    this.hooks.onEnd?.(msg); break
      case "teleport":
      case "switch":
        if (msg.ok === false) this.flash({ cooldown: `Actie beschikbaar over ${this.countdown()}`, bounds: "Te ver van de arena", status: "Geen ronde bezig" }[msg.reason] ?? "Dat gaat niet")
        else if (msg.id === this.playerId) { this.nextActionAt = msg.next_action_at; this.hooks.onAction?.(msg) }
        break
    }
  }

  setRound(body, live) {
    const prev = this.round
    this.round = body
    this.damage = body?.damage ?? 0                                  // a new town starts the bill again
    this.obstacles = new Map((body?.obstacles ?? []).map((o) => [o.key, o]))
    if (body) {
      const fresh = body.id !== prev?.id, started = body.status === "running" && prev?.status !== "running"
      if (fresh || started) this.nextActionAt = null                 // everyone's action is ready for a new town
      this.hooks.onRound?.(body, { fresh, started, live })
    }
    this.hud()
    this.showDamage()
  }

  // free roam has no server to keep the total, so it keeps its own
  addDamage(euros) {
    if (!(euros > 0)) return
    this.damage += euros
    this.showDamage()
  }

  showDamage() {
    const el = this.els.schade
    if (!el) return
    el.hidden = !(this.damage > 0)
    if (!el.hidden) el.textContent = `${euro(this.damage)} schade`
  }

  setVote(vote) {
    if (!vote && !this.vote) return
    this.vote = vote
    this.hooks.onVote?.(vote)
  }

  applyObjects(list, damage) {
    if (damage != null) this.damage = damage
    for (const o of list) {
      const ob = this.obstacles.get(o.key)
      if (ob) { ob.state = o.state; ob.hp = o.hp; ob.max = o.max }
    }
    this.hooks.onObjects?.(list)
    this.showDamage()
  }

  get remaining() {
    let n = 0
    for (const o of this.obstacles.values()) if (o.state !== "gone") n++
    return n
  }

  // the float's motion, mirroring Game::Round on the server
  travelled(now = this.now()) {
    const r = this.round
    if (!r?.started_at) return 0
    return THREE.MathUtils.clamp(r.speed * (now - r.started_at) / 1000, 0, r.path.length)
  }

  progress(now) { return this.round ? this.travelled(now) / this.round.path.length : 0 }

  // How far along the parade route a point lies, or null if the float does not have to care about it: off to the
  // side, behind it, or so close in front that nobody could clear it in time. The server applies the same rule, so
  // this only saves the traffic.
  onRoute(x, z) {
    const r = this.round
    if (!this.running || !r) return null
    const p = r.path, dx = p.x1 - p.x0, dz = p.z1 - p.z0
    const len2 = dx * dx + dz * dz || 1
    const t = ((x - p.x0) * dx + (z - p.z0) * dz) / len2
    const at = t * p.length
    if (at <= this.travelled() + PARADE.ahead || at >= p.length) return null
    const off = Math.hypot(x - (p.x0 + dx * t), z - (p.z0 + dz * t))
    return off <= PARADE.corridor ? at : null
  }

  floatAt(now) {
    const p = this.round.path, t = this.travelled(now) / p.length
    return [p.x0 + (p.x1 - p.x0) * t, p.z0 + (p.z1 - p.z0) * t]
  }

  get heading() { const p = this.round.path; return Math.atan2(-(p.x1 - p.x0), -(p.z1 - p.z0)) }   // yaw: 0 = north, positive = left

  canAct() { return !this.nextActionAt || this.nextActionAt <= this.now() }

  countdown(at = this.nextActionAt) {
    const s = Math.max(0, Math.ceil((at - this.now()) / 1000))
    return `${Math.floor(s / 60)}:${String(s % 60).padStart(2, "0")}`
  }

  // every 0.25 s from game.js
  hud() {
    const { actie, banner, route } = this.els
    const r = this.round
    actie.hidden = !r
    if (!r) { banner.hidden = route.hidden = true; return }
    const secs = (at) => Math.max(0, Math.ceil((at - this.now()) / 1000))
    const naam = r.arena.name
    if (r.status === "running") {
      banner.hidden = true
      route.hidden = false
      this.routeBar()
    } else if (r.status === "intermission") {
      route.hidden = true
      this.showBanner(naam, `Optocht start over ${secs(r.next_at)} s`)
    } else {
      route.hidden = true
      this.showBanner(r.result === "won" ? "Alaaf! Optocht binnen" : "Optocht vastgelopen", `Volgende plaats over ${secs(r.next_at)} s`)
    }
    actie.textContent = this.canAct() ? "Actie: klaar" : `Actie over ${this.countdown()}`
  }

  // the route as a vertical bar, start at the bottom: the float where it is, an icon per stretch of route for the
  // obstacles still standing there (the commonest kind, with a count when there are more), the cleared part tinted
  // behind the float. Rendered into the HUD's bar during the round and into the loading screen's before it.
  routeBar(container = this.els.route) {
    const r = this.round, len = r.path.length, p = this.progress()
    const q = (cls) => container.querySelector(cls)
    q(".route-kop").textContent = `Ronde ${r.id} · ${r.arena.name}`
    q(".route-gedaan").style.height = q(".route-optocht").style.bottom = `${(p * 100).toFixed(1)}%`
    const buckets = new Map()
    for (const o of this.obstacles.values()) {
      if (o.state === "gone") continue
      const i = THREE.MathUtils.clamp(Math.floor(o.at / len * SLOTS), 0, SLOTS - 1)
      const b = buckets.get(i) ?? { n: 0, rubble: 0, kinds: {} }
      b.n++; b.kinds[o.kind] = (b.kinds[o.kind] ?? 0) + 1
      if (o.state === "rubble" || o.kind === "d") b.rubble++
      buckets.set(i, b)
    }
    const html = [...buckets].sort((a, b) => a[0] - b[0]).map(([i, b]) => {
      const kind = Object.entries(b.kinds).sort((a, c) => c[1] - a[1])[0][0]
      const icon = kind === "d" || b.rubble >= b.n ? RUBBLE_ICON : ICON[kind] ?? ICON.m
      return `<span class="route-icoon${b.n > 1 ? " meer" : ""}" style="bottom:${((i + 0.5) / SLOTS * 100).toFixed(1)}%" data-n="${b.n}">${icon}</span>`
    }).join("")
    if (html !== container.routeHtml) { container.routeHtml = html; q(".route-iconen").innerHTML = html }
  }

  showBanner(title, sub) {
    const { banner, bannerTitel, bannerSub } = this.els
    bannerTitel.textContent = title
    bannerSub.textContent = sub
    banner.hidden = false
  }

  // a short toast at the bottom of the screen
  flash(text) {
    const { flits } = this.els
    flits.textContent = text
    flits.hidden = false
    clearTimeout(this.flashTimer)
    this.flashTimer = setTimeout(() => { flits.hidden = true }, 2200)
  }
}

File: lib/game/round.rb
module Game
  # One round: the arena, the parade float's motion along the route, every object hit so far and each player's
  # action cooldown. Time is server milliseconds. The float's position is a pure function of started_at and speed,
  # so clients render it from a clock offset without any messages.
  class Round
    CROSSING_MS    = 900_000                      # the float takes 15 minutes edge to edge
    RUBBLE_SHARE   = 0.5                          # a flattened building leaves rubble worth this share of its hit points
    ACTION_MS      = 60_000                       # teleport or vehicle switch, then this long a cooldown
    MAX_HP         = 100_000
    FLOAT          = { length: 12, width: 4 }.freeze
    BUILDING_KINDS = %w[m b].freeze

    MAX_WOZ        = 50_000_000                   # € a client may claim one building is worth (lib/tasks/woz.rake caps it there)

    # Debris in the road. A heap covers this much of the route, costs this much to sweep aside, and may only be
    # reported this far ahead of the float — anything nearer would be a heap the parade has no time to get past,
    # and anything behind it has already been driven through.
    DEBRIS_SLOT    = 8.0
    DEBRIS_HP      = 30
    DEBRIS_AHEAD   = 40.0
    DEBRIS_MAX     = 60                           # heaps at once, so a demolition derby cannot fill the route

    Obj    = Struct.new(:key, :kind, :x, :z, :at, :hp, :max, :state, :woz, keyword_init: true)   # at: nil for objects off the path
    Player = Struct.new(:id, :name, :vehicle, :joined_at, :last_action_at, :x, :z, :tabs, keyword_init: true)

    attr_reader :id, :status, :result, :arena, :path, :spawn, :speed, :started_at, :ended_at, :obstacles, :objects, :damage
    attr_accessor :next_at

    def initialize(id:, arena:, path:, spawn:, obstacles:, next_at: nil)
      @id, @arena, @path, @spawn, @next_at = id, arena, path, spawn, next_at
      @speed = path[:length] / (CROSSING_MS / 1000.0)
      @obstacles = obstacles.map { Obj.new(**_1, state: :intact) }.sort_by(&:at)
      @objects = @obstacles.to_h { [ _1.key, _1 ] }        # everything touched this round, corridor objects included
      @damage = 0.0                                        # euros of property flattened, everybody's, this round
      @status = :intermission
    end

    def start!(now, players)
      @status, @started_at, @next_at = :running, now, nil
      players.each { _1.last_action_at = nil }
    end

    def end!(now, result)
      @status, @result, @ended_at = :ended, result, now
    end

    def running? = status == :running
    def length = path[:length]
    def ends_at = started_at && started_at + CROSSING_MS
    def travelled(now) = ((now - started_at) / 1000.0 * speed).clamp(0.0, length)

    def float_at(now)
      t = travelled(now) / length
      [ path[:x0] + (path[:x1] - path[:x0]) * t, path[:z0] + (path[:z1] - path[:z0]) * t ]
    end

    def blocker = obstacles.find { _1.state != :gone }

    # A heap of rubble has come to rest in the road. The clients report where their own debris settled; the route is
    # bucketed so a house that sheds fifty pieces over eight metres becomes one thing to clear rather than fifty,
    # and it only counts ahead of the float, since rubble behind it is rubble the parade has already passed.
    def debris(at, now)
      return unless running?
      ahead = travelled(now) + DEBRIS_AHEAD
      return unless at > ahead && at < length
      slot = (at / DEBRIS_SLOT).floor
      key = "d:#{slot}"
      if (obj = objects[key])
        return if obj.state == :gone
        obj.hp = [ obj.hp + DEBRIS_HP, MAX_HP ].min
        obj.max = [ obj.max, obj.hp ].max
        return obj
      end
      return if objects.count { |_, o| o.kind == "d" } >= DEBRIS_MAX
      mid = slot * DEBRIS_SLOT + DEBRIS_SLOT / 2
      x, z = point_at(mid)
      obj = Obj.new(key:, kind: "d", x: x.round(2), z: z.round(2), at: mid, hp: DEBRIS_HP, max: DEBRIS_HP, state: :intact)
      objects[key] = obj
      i = obstacles.index { _1.at > mid } || obstacles.size      # the list is kept in route order; blocker walks it
      obstacles.insert(i, obj)
      obj
    end

    # a point this far along the route
    def point_at(d)
      t = (d / length).clamp(0.0, 1.0)
      [ path[:x0] + (path[:x1] - path[:x0]) * t, path[:z0] + (path[:z1] - path[:z0]) * t ]
    end

    # :lost when the float's nose reaches something still standing, :won when it reaches the far edge
    def check(now)
      d = travelled(now)
      if (b = blocker) && d >= b.at then :lost
      elsif d >= length then :won
      end
    end

    # One damage report. `max` is pinned when the object is first seen: clients compute it from the object's size,
    # the server does not know sizes. Buildings crumble to rubble at zero and get RUBBLE_SHARE of their hit points
    # back; everything else is gone at once. Returns the object when it changed, nil otherwise.
    def hit(key, damage, max, woz = nil)
      return if damage <= 0
      obj = objects[key] ||= Obj.new(key:, kind: key[0], state: :intact)
      return if obj.state == :gone
      obj.max ||= max.clamp(1, MAX_HP)
      obj.hp ||= obj.max
      obj.woz ||= woz&.clamp(0, MAX_WOZ)
      # The bill, in euros: what this hit actually took off, as a share of the whole. Whoever reports the first hit
      # on a building says what it is worth and nobody can raise it afterwards, and only an intact one can be
      # charged for — so a building adds its own value to the total once and clearing its rubble afterwards is free,
      # however many players are hammering it.
      @damage += obj.woz * [ damage, obj.hp ].min / obj.max if obj.woz && obj.hp.positive? && obj.state == :intact
      obj.hp -= damage
      if obj.hp <= 0 && BUILDING_KINDS.include?(obj.kind) && obj.state == :intact
        obj.state, obj.hp = :rubble, (obj.max * RUBBLE_SHARE).ceil
      elsif obj.hp <= 0
        obj.state, obj.hp = :gone, 0
      end
      obj
    end

    def action_allowed?(player, now) = player.last_action_at.nil? || now - player.last_action_at >= ACTION_MS
    def next_action_at(player) = player.last_action_at && player.last_action_at + ACTION_MS

    def inside_arena?(x, z, margin = 500)
      (x - arena[:cx]).abs <= arena[:half] + margin && (z - arena[:cz]).abs <= arena[:half] + margin
    end

    def to_h(players)
      { id:, status:, result:, arena:, path:, speed: speed.round(4), float: FLOAT, damage: damage.round,
        started_at:, ends_at:, next_at:, spawn:,
        obstacles: obstacles.map { obj_h(_1).merge(kind: _1.kind, x: _1.x, z: _1.z, at: _1.at) },
        objects: objects.values.reject(&:at).map { obj_h(_1) },
        players: players.map { { id: _1.id, name: _1.name, vehicle: _1.vehicle } } }
    end

    def obj_h(o) = { key: o.key, hp: o.hp&.round(1), max: o.max, state: o.state }
  end
end
